File: test_chart_mqtt.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
MqttPanelCraft - 折線圖 (CHART) 元件 MQTT 測試腳本
支援單線數值傳輸與多線序列 (JSON 格式) 傳輸

相依套件安裝：
    pip install paho-mqtt
"""

import time
import math
import json
import argparse
import sys

try:
    import paho.mqtt.client as mqtt
except ImportError:
    print("請先安裝 paho-mqtt 套件： pip install paho-mqtt")
    sys.exit(1)

def main():
    parser = argparse.ArgumentParser(description="MqttPanelCraft 折線圖元件測試發送工具")
    parser.add_argument("--broker", default="broker.emqx.io", help="MQTT Broker 地址 (預設: broker.emqx.io)")
    parser.add_argument("--port", type=int, default=1883, help="MQTT Broker 埠號 (預設: 1883)")
    parser.add_argument("--topic", default="home/chart/data", help="目標 CHART 元件訂閱的 Topic (預設: home/chart/data)")
    parser.add_argument("--mode", choices=["single", "multi"], default="multi", help="測試模式：single (單一數字) 或 multi (JSON 多序列，例如 value 與 series2)")
    parser.add_argument("--interval", type=float, default=1.0, help="發送時間間隔秒數 (預設: 1.0 秒)")
    args = parser.parse_args()

    print(f"[*] 正在連線至 MQTT Broker: {args.broker}:{args.port} ...")
    client = mqtt.Client()
    try:
        client.connect(args.broker, args.port, 60)
    except Exception as e:
        print(f"[!] 連線失敗: {e}")
        sys.exit(1)

    client.loop_start()
    print(f"[+] 連線成功！準備持續發送測試數據至 Topic: {args.topic}")
    print(f"[*] 測試模式: {args.mode.upper()} (按 Ctrl+C 可停止發送)")

    step = 0
    try:
        while True:
            # 使用正弦波與餘弦波模擬真實物聯網動態數據 (如溫度、濕度、電壓波動)
            val1 = round(50 + 30 * math.sin(step * 0.3) + (step % 5), 1)
            val2 = round(40 + 25 * math.cos(step * 0.25), 1)
            val3 = round(60 + 15 * math.sin(step * 0.5), 1)

            if args.mode == "single":
                # 單純發送單浮點數 (對應單線序列)
                payload = str(val1)
            else:
                # 發送 JSON 格式多序列數據 (預設對應 series_key_1=value, series_key_2=series2, series_key_3=series3)
                payload = json.dumps({
                    "value": val1,
                    "series2": val2,
                    "series3": val3,
                    "t": time.strftime("%H:%M:%S")
                })

            client.publish(args.topic, payload, qos=0)
            print(f"[{time.strftime('%H:%M:%S')}] 已發送至 {args.topic} -> {payload}")
            
            step += 1
            time.sleep(args.interval)
    except KeyboardInterrupt:
        print("\n[*] 接收到停止指令，關閉連線中...")
    finally:
        client.loop_stop()
        client.disconnect()
        print("[+] 測試腳本已結束")

if __name__ == "__main__":
    main()

File: test_chart_mqtt.js
/**
 * MqttPanelCraft - 折線圖 (CHART) 元件 MQTT Node.js 測試腳本
 * 由於系統中未檢測到 Python 執行環境，本腳本使用 Node.js 進行測試。
 * 支援單線數值傳輸與多線序列 (JSON 格式) 傳輸。
 *
 * 使用說明與相依安裝：
 *   npm install mqtt minimist
 *   node test_chart_mqtt.js --topic "home/chart/data" --mode multi
 */

const mqtt = require('mqtt');

// 解析命令列參數
const args = process.argv.slice(2);
function getArg(name, defaultValue) {
    const idx = args.indexOf(`--${name}`);
    if (idx !== -1 && idx + 1 < args.length) {
        return args[idx + 1];
    }
    return defaultValue;
}

const broker = getArg('broker', 'broker.emqx.io');
const port = parseInt(getArg('port', '1883'), 10);
const topic = getArg('topic', 'home/chart/data');
const mode = getArg('mode', 'multi'); // 'single' 或 'multi'
const interval = parseFloat(getArg('interval', '1.0')) * 1000;

const brokerUrl = `mqtt://${broker}:${port}`;
console.log(`[*] 正在連線至 MQTT Broker: ${brokerUrl} ...`);

const client = mqtt.connect(brokerUrl);

client.on('connect', () => {
    console.log(`[+] 連線成功！準備持續發送測試數據至 Topic: ${topic}`);
    console.log(`[*] 測試模式: ${mode.toUpperCase()} (按 Ctrl+C 可停止發送)\n`);

    let step = 0;
    setInterval(() => {
        // 使用正弦波與餘弦波模擬真實物聯網動態數據 (如溫度、濕度、電壓波動)
        const val1 = Math.round((50 + 30 * Math.sin(step * 0.3) + (step % 5)) * 10) / 10;
        const val2 = Math.round((40 + 25 * Math.cos(step * 0.25)) * 10) / 10;
        const val3 = Math.round((60 + 15 * Math.sin(step * 0.5)) * 10) / 10;

        let payload = '';
        if (mode === 'single') {
            payload = String(val1);
        } else {
            const now = new Date();
            const timeStr = now.toTimeString().split(' ')[0];
            payload = JSON.stringify({
                value: val1,
                series2: val2,
                series3: val3,
                t: timeStr
            });
        }

        client.publish(topic, payload, { qos: 0 });
        const timeLog = new Date().toTimeString().split(' ')[0];
        console.log(`[${timeLog}] 已發送至 ${topic} -> ${payload}`);

        step++;
    }, interval);
});

client.on('error', (err) => {
    console.error(`[!] MQTT 連線錯誤: ${err.message}`);
    process.exit(1);
});
